### DISCUSSION TITLE
[@types/node] `fs.createWriteStream` constructor is incorrectly typed.

### DISCUSSION THREAD

**Discussion by Gadzy** [DefinitelyTyped/DefinitelyTyped#65673] [Issues with a @types package]
I'm aware that `fs.WriteStream` should not be directly instantiated, but rather created by calling `fs.createWriteStream`, though I'm not sure why. At first I thought it was because - since Node v12.x if I'm correct - [some lazy loading happens in that function](https://github.com/nodejs/node/blob/v20.x/lib/fs.js#L3067), until I discovered that [a getter ensures the lazy loading is done anyway](https://github.com/nodejs/node/blob/v20.x/lib/fs.js#L3174).

Still, it is possible to do so, but as `@types/node` is missing the constructor signature for `fs.WriteStream`, TypeScript believes the one from it's super class `stream.Writable` applies, which is wrong.

The correct constructor signature should be the same as `fs.createWriteStream` (see [here](https://github.com/nodejs/node/blob/v20.x/lib/fs.js#L3068) and [there](https://github.com/nodejs/node/blob/v20.x/lib/internal/fs/streams.js#L307)) :
```ts
constructor(path: PathLike, options?: BufferEncoding | StreamOptions): WriteStream;
```

I can provide a pull request if needed, the addition is pretty straightforward.

[a comment by an automated account is not shown]

**Comment by peterblazejewicz** [comment]
This is internal implementation, hidden behind factory methods. Public documentation has no creational section at all:
https://nodejs.org/api/fs.html#class-fswritestream
Why would you want to have this exposed/leaked?

**Comment by Gadzy** [comment]
Thanks for pointing at the public documentation.
Couldn't it be typed as an interface rather than a class to enforce use of the factory method ?

The motivation behind this issue is that the following compiles but throws a runtime error that is hard to understand given that Intelllisense doesn't mention this `path` argument :
```ts
// example.ts
import { WriteStream } from "node:fs";
new WriteStream();
```
```
$ npx ts-node example.ts
TypeError [ERR_INVALID_ARG_TYPE]: The "path" argument must be of type string or an instance of Buffer or URL. Received undefined
```

The problem arose while unit-testing a service that makes inner calls to `createWriteStream`.
We wanted to mock those inner calls return value with an error-emitting subclass of `WriteStream` so we could test the case when write access was denied.
